src/freenet/pluginmanager/FredPluginFCPMessageHandler.java

```diff
@@ -0,0 +1,580 @@
+         *             explains how you can nevertheless send a reply to reply messages.
+         */
+        public static FCPPluginMessage constructReplyMessage(FCPPluginMessage originalMessage,
+                SimpleFieldSet params, Bucket data, boolean success, String errorCode,
+                String errorMessage) {
+            
+            if(originalMessage.isReplyMessage()) {
+                throw new IllegalStateException("Constructing a reply message for a message which "
+                    + "was a reply message already not allowed.");
+            }
+            
+            return new FCPPluginMessage(null, originalMessage.identifier,
+                params, data, success, errorCode, errorMessage);
+        }
+
+        /**
+         * Same as {@link #constructReplyMessage(FCPPluginMessage, SimpleFieldSet, Bucket, boolean,
+         * String, String)} with the missing parameters being:<br>
+         * <code>
+         * SimpleFieldSet params = new SimpleFieldSet(shortLived = true);<br>
+         * Bucket data = null;<br>
+         * boolean success = true;<br>
+         * errorCode = null;<br>
+         * errorMessage = null;<br>
+         * </code>
+         */
+        public static FCPPluginMessage constructSuccessReply(FCPPluginMessage originalMessage) {
+            return constructReplyMessage(
+                originalMessage, new SimpleFieldSet(true), null, true, null, null);
+        }
+
+        /**
+         * Same as {@link #constructReplyMessage(FCPPluginMessage, SimpleFieldSet, Bucket, boolean,
+         * String, String)} with the missing parameters being:<br>
+         * <code>
+         * SimpleFieldSet params = new SimpleFieldSet(shortLived = true);<br>
+         * Bucket data = null;<br>
+         * boolean success = false;<br>
+         * </code>
+         */
+        public static FCPPluginMessage constructErrorReply(FCPPluginMessage originalMessage,
+                String errorCode, String errorMessage) {
+            
+            return constructReplyMessage(
+                originalMessage, new SimpleFieldSet(true), null, false, errorCode, errorMessage);
+        }
+
+        /**
+         * ATTENTION: Only for being used by internal network code.<br><br>
+         * 
+         * You <b>must not</b> use this for constructing outgoing messages in server or client
+         * implementations.<br>
+         * 
+         * This function is typically to construct incoming messages for passing them to the message
+         * handling function
+         * {@link FredPluginFCPMessageHandler#handlePluginFCPMessage(FCPPluginClient,
+         * FCPPluginMessage)}.<br><br>
+         * 
+         * See the JavaDoc of the member variables with the same name as the parameters for an
+         * explanation of the parameters.<br>
+         */
+        public static FCPPluginMessage constructRawMessage(ClientPermissions permissions,
```

Comment:
TODO for myself: This cannot be package private at the moment since the code which uses it is in package freenet.clients.fcp. However, one we move this member class FCPPluginMessage to its own file in package freenet.clients.fcp, it can be. It has grown large enough for this to make sense anyway.